File: rust-webgl/lib/src/glmath/matrix4.rs
use std::ops::Mul;

use super::{
    angles::Radians,
    numbers::{CouldBeAnAngle, Float},
    vector3::Vector3,
};

#[repr(C)]
#[derive(Debug, Clone, Copy)]
pub struct Matrix4<T> {
    data: [[T; 4]; 4],
}

impl<T> Matrix4<T> {
    pub fn flatten(&self) -> &[T] {
        self.data.flatten()
    }
}

impl<T> Matrix4<T>
where
    T: Float + Copy,
    Radians<T>: CouldBeAnAngle<Output = T>,
{
    pub fn new_identity() -> Self {
        Self {
            data: [
                [T::ONE, T::ZERO, T::ZERO, T::ZERO],
                [T::ZERO, T::ONE, T::ZERO, T::ZERO],
                [T::ZERO, T::ZERO, T::ONE, T::ZERO],
                [T::ZERO, T::ZERO, T::ZERO, T::ONE],
            ],
        }
    }

    pub fn new_translation(v: Vector3<T>) -> Self {
        Self {
            data: [
                [T::ONE, T::ZERO, T::ZERO, T::ZERO],
                [T::ZERO, T::ONE, T::ZERO, T::ZERO],
                [T::ZERO, T::ZERO, T::ONE, T::ZERO],
                [v.x, v.y, v.z, T::ONE],
            ],
        }
    }

    pub fn new_scale(v: Vector3<T>) -> Self {
        Self {
            data: [
                [v.x, T::ZERO, T::ZERO, T::ZERO],
                [T::ZERO, v.y, T::ZERO, T::ZERO],
                [T::ZERO, T::ZERO, v.z, T::ZERO],
                [T::ZERO, T::ZERO, T::ZERO, T::ONE],
            ],
        }
    }

    pub fn new_rotation(angle: Radians<T>, axis: Vector3<T>) -> Self {
        let c: T = angle.cos();
        let s: T = angle.sin();
        let one_minus_c = T::ONE - c;
        let axis = axis.normalized();
        Self {
            data: [
                [
                    axis.x * axis.x * one_minus_c + c,
                    axis.x * axis.y * one_minus_c - axis.z * s,
                    axis.x * axis.z * one_minus_c + axis.y * s,
                    T::ZERO,
                ],
                [
                    axis.y * axis.x * one_minus_c + axis.z * s,
                    axis.y * axis.y * one_minus_c + c,
                    axis.y * axis.z * one_minus_c - axis.x * s,
                    T::ZERO,
                ],
                [
                    axis.x * axis.z * one_minus_c - axis.y * s,
                    axis.y * axis.z * one_minus_c + axis.x * s,
                    axis.z * axis.z * one_minus_c + c,
                    T::ZERO,
                ],
                [T::ZERO, T::ZERO, T::ZERO, T::ONE],
            ],
        }
    }

    pub fn new_ortho(left: T, right: T, bottom: T, top: T, near: T, far: T) -> Self {
        Self {
            data: [
                [T::TWO / (right - left), T::ZERO, T::ZERO, T::ZERO],
                [T::ZERO, T::TWO / (top - bottom), T::ZERO, T::ZERO],
                [T::ZERO, T::ZERO, -T::TWO / (far - near), T::ZERO],
                [
                    -(right + left) / (right - left),
                    -(top + bottom) / (top - bottom),
                    -(far + near) / (far - near),
                    T::ONE,
                ],
            ],
        }
    }

    pub fn new_perspective(fov: Radians<T>, width: T, height: T, near: T, far: T) -> Self {
        let f = T::ONE / (fov.0 / T::TWO).tan();
        let aspect = width / height;
        Self {
            data: [
                [f / aspect, T::ZERO, T::ZERO, T::ZERO],
                [T::ZERO, f, T::ZERO, T::ZERO],
                [T::ZERO, T::ZERO, (far + near) / (near - far), -T::ONE],
                [T::ZERO, T::ZERO, T::TWO * far * near / (near - far), T::ONE],
            ],
        }
    }

    pub fn new_look_at(position: Vector3<T>, target: Vector3<T>, up: Vector3<T>) -> Self {
        let f = (target - position).normalized();
        let up = up.normalized();
        let s = f.cross_product(up).normalized();
        let u = s.cross_product(f).normalized();
        *Self {
            data: [
                [s.x, u.x, -f.x, T::ZERO],
                [s.y, u.y, -f.y, T::ZERO],
                [s.z, u.z, -f.z, T::ZERO],
                [T::ZERO, T::ZERO, T::ZERO, T::ONE],
            ],
        }
        .translate(-position)
    }

    pub fn append(&mut self, other: Self) -> &mut Self {
        let other = other * *self;
        self.data = other.data;
        self
    }

    pub fn translate(&mut self, v: Vector3<T>) -> &mut Self {
        self.append(Matrix4::new_translation(v))
    }

    pub fn scale(&mut self, v: Vector3<T>) -> &mut Self {
        self.append(Matrix4::new_scale(v))
    }

    pub fn rotate(&mut self, angle: Radians<T>, axis: Vector3<T>) -> &mut Self {
        self.append(Matrix4::new_rotation(angle, axis))
    }
}

impl<T> Matrix4<T>
where
    T: Float + Copy,
{
    pub fn apply_to_point(&self, input: Vector3<T>) -> Vector3<T> {
        Vector3::new(
            self.data[0][0] * input.x
                + self.data[1][0] * input.y
                + self.data[2][0] * input.z
                + self.data[3][0],
            self.data[0][1] * input.x
                + self.data[1][1] * input.y
                + self.data[2][1] * input.z
                + self.data[3][1],
            self.data[0][2] * input.x
                + self.data[1][2] * input.y
                + self.data[2][2] * input.z
                + self.data[3][2],
        )
    }

    pub fn apply_to_vector(&self, input: Vector3<T>) -> Vector3<T> {
        Vector3::new(
            self.data[0][0] * input.x + self.data[1][0] * input.y + self.data[2][0] * input.z,
            self.data[0][1] * input.x + self.data[1][1] * input.y + self.data[2][1] * input.z,
            self.data[0][2] * input.x + self.data[1][2] * input.y + self.data[2][2] * input.z,
        )
    }
}

impl<T> Mul for Matrix4<T>
where
    T: Float + Copy,
{
    type Output = Matrix4<T>;

    fn mul(self, rhs: Self) -> Self::Output {
        let mut data = [
            [T::ZERO, T::ZERO, T::ZERO, T::ZERO],
            [T::ZERO, T::ZERO, T::ZERO, T::ZERO],
            [T::ZERO, T::ZERO, T::ZERO, T::ZERO],
            [T::ZERO, T::ZERO, T::ZERO, T::ZERO],
        ];
        for (i, row) in data.iter_mut().enumerate() {
            for (j, col) in row.iter_mut().enumerate() {
                for k in 0..4 {
                    *col += self.data[i][k] * rhs.data[k][j];
                }
            }
        }
        Self::Output { data }
    }
}

#[cfg(test)]
pub mod tests {
    use std::fmt::Display;

    use crate::glmath::angles::Degrees;

    use super::*;

    #[test]
    fn identity() {
        let m = Matrix4::new_identity();
        assert_vector3_close_to(
            m.apply_to_point(Vector3::new(1f64, 2f64, 3f64)),
            Vector3::new(1f64, 2f64, 3f64),
            1e-8f64,
        );
        assert_vector3_close_to(
            m.apply_to_vector(Vector3::new(1f64, 2f64, 3f64)),
            Vector3::new(1f64, 2f64, 3f64),
            1e-8f64,
        );
    }

    #[test]
    fn translation() {
        let m = Matrix4::new_translation(Vector3::new(1f64, 2f64, 3f64));
        assert_vector3_close_to(
            m.apply_to_point(Vector3::new(1f64, 1f64, 1f64)),
            Vector3::new(2f64, 3f64, 4f64),
            1e-8f64,
        );
        assert_vector3_close_to(
            m.apply_to_vector(Vector3::new(1f64, 1f64, 1f64)),
            Vector3::new(1f64, 1f64, 1f64),
            1e-8f64,
        );
    }

    #[test]
    fn scale() {
        let m = Matrix4::new_scale(Vector3::new(2f64, 3f64, 4f64));
        assert_vector3_close_to(
            m.apply_to_point(Vector3::new(2f64, 2f64, 2f64)),
            Vector3::new(4f64, 6f64, 8f64),
            1e-8f64,
        );
        assert_vector3_close_to(
            m.apply_to_vector(Vector3::new(2f64, 2f64, 2f64)),
            Vector3::new(4f64, 6f64, 8f64),
            1e-8f64,
        );
    }

    #[test]
    fn rotation_x() {
        let m = Matrix4::new_rotation(Degrees(90f64).into(), Vector3::new(1f64, 0f64, 0f64));
        assert_vector3_close_to(
            m.apply_to_point(Vector3::new(1f64, 1f64, 1f64)),
            Vector3::new(1f64, 1f64, -1f64),
            1e-8f64,
        );
        assert_vector3_close_to(
            m.apply_to_vector(Vector3::new(1f64, 1f64, 1f64)),
            Vector3::new(1f64, 1f64, -1f64),
            1e-8f64,
        );

        let m = Matrix4::new_rotation(Degrees(-90f64).into(), Vector3::new(1f64, 0f64, 0f64));
        assert_vector3_close_to(
            m.apply_to_point(Vector3::new(1f64, 1f64, 1f64)),
            Vector3::new(1f64, -1f64, 1f64),
            1e-8f64,
        );
        assert_vector3_close_to(
            m.apply_to_vector(Vector3::new(1f64, 1f64, 1f64)),
            Vector3::new(1f64, -1f64, 1f64),
            1e-8f64,
        );

        let m = Matrix4::new_rotation(Degrees(90f64).into(), Vector3::new(-1f64, 0f64, 0f64));
        assert_vector3_close_to(
            m.apply_to_point(Vector3::new(1f64, 1f64, 1f64)),
            Vector3::new(1f64, -1f64, 1f64),
            1e-8f64,
        );
        assert_vector3_close_to(
            m.apply_to_vector(Vector3::new(1f64, 1f64, 1f64)),
            Vector3::new(1f64, -1f64, 1f64),
            1e-8f64,
        );
    }

    #[test]
    fn rotation_y() {
        let m = Matrix4::new_rotation(Degrees(90f64).into(), Vector3::new(0f64, 1f64, 0f64));
        assert_vector3_close_to(
            m.apply_to_point(Vector3::new(1f64, 1f64, 1f64)),
            Vector3::new(-1f64, 1f64, 1f64),
            1e-8f64,
        );
        assert_vector3_close_to(
            m.apply_to_vector(Vector3::new(1f64, 1f64, 1f64)),
            Vector3::new(-1f64, 1f64, 1f64),
            1e-8f64,
        );

        let m = Matrix4::new_rotation(Degrees(-90f64).into(), Vector3::new(0f64, 1f64, 0f64));
        assert_vector3_close_to(
            m.apply_to_point(Vector3::new(1f64, 1f64, 1f64)),
            Vector3::new(1f64, 1f64, -1f64),
            1e-8f64,
        );
        assert_vector3_close_to(
            m.apply_to_vector(Vector3::new(1f64, 1f64, 1f64)),
            Vector3::new(1f64, 1f64, -1f64),
            1e-8f64,
        );

        let m = Matrix4::new_rotation(Degrees(90f64).into(), Vector3::new(0f64, -1f64, 0f64));
        assert_vector3_close_to(
            m.apply_to_point(Vector3::new(1f64, 1f64, 1f64)),
            Vector3::new(1f64, 1f64, -1f64),
            1e-8f64,
        );
        assert_vector3_close_to(
            m.apply_to_vector(Vector3::new(1f64, 1f64, 1f64)),
            Vector3::new(1f64, 1f64, -1f64),
            1e-8f64,
        );
    }

    #[test]
    fn rotation_z() {
        let m = Matrix4::new_rotation(Degrees(90f64).into(), Vector3::new(0f64, 0f64, 1f64));
        assert_vector3_close_to(
            m.apply_to_point(Vector3::new(1f64, 1f64, 1f64)),
            Vector3::new(1f64, -1f64, 1f64),
            1e-8f64,
        );
        assert_vector3_close_to(
            m.apply_to_vector(Vector3::new(1f64, 1f64, 1f64)),
            Vector3::new(1f64, -1f64, 1f64),
            1e-8f64,
        );

        let m = Matrix4::new_rotation(Degrees(-90f64).into(), Vector3::new(0f64, 0f64, 1f64));
        assert_vector3_close_to(
            m.apply_to_point(Vector3::new(1f64, 1f64, 1f64)),
            Vector3::new(-1f64, 1f64, 1f64),
            1e-8f64,
        );
        assert_vector3_close_to(
            m.apply_to_vector(Vector3::new(1f64, 1f64, 1f64)),
            Vector3::new(-1f64, 1f64, 1f64),
            1e-8f64,
        );

        let m = Matrix4::new_rotation(Degrees(90f64).into(), Vector3::new(0f64, 0f64, -1f64));
        assert_vector3_close_to(
            m.apply_to_point(Vector3::new(1f64, 1f64, 1f64)),
            Vector3::new(-1f64, 1f64, 1f64),
            1e-8f64,
        );
        assert_vector3_close_to(
            m.apply_to_vector(Vector3::new(1f64, 1f64, 1f64)),
            Vector3::new(-1f64, 1f64, 1f64),
            1e-8f64,
        );
    }

    #[test]
    fn translation_and_rotation() {
        let m = *Matrix4::new_identity()
            .rotate(Degrees(90f64).into(), Vector3::new(0f64, 1f64, 0f64))
            .translate(Vector3::new(1f64, 2f64, 3f64));
        assert_vector3_close_to(
            m.apply_to_point(Vector3::new(1f64, 1f64, 1f64)),
            Vector3::new(-4f64, 3f64, 2f64),
            1e-8f64,
        );
        assert_vector3_close_to(
            m.apply_to_vector(Vector3::new(1f64, 1f64, 1f64)),
            Vector3::new(-1f64, 1f64, 1f64),
            1e-8f64,
        );
    }

    fn assert_vector3_close_to<T>(a: Vector3<T>, b: Vector3<T>, max_distance: T)
    where
        T: Float + Copy + PartialOrd + Display,
    {
        let actual_distance = (a - b).magnitude();
        assert!(
            actual_distance < max_distance,
            "expected {} and {} to be within {}, actual distance {}",
            a,
            b,
            max_distance,
            actual_distance
        );
    }
}

File: swift-c-interop-sdl3-wgpu/Sources/Experiment/math-and-graphics-types/Matrix4.swift
struct Matrix4<T> {
    let m00: T, m01: T, m02: T, m03: T
    let m10: T, m11: T, m12: T, m13: T
    let m20: T, m21: T, m22: T, m23: T
    let m30: T, m31: T, m32: T, m33: T
}

extension Matrix4 where T: ExpressibleByIntegerLiteral {
    static var identity: Self {
        Self(
            m00: 1, m01: 0, m02: 0, m03: 0,
            m10: 0, m11: 1, m12: 0, m13: 0,
            m20: 0, m21: 0, m22: 1, m23: 0,
            m30: 0, m31: 0, m32: 0, m33: 1
        )
    }

    static func translation(_ v: Vector3<T>) -> Self {
        Self(
            m00: 1, m01: 0, m02: 0, m03: 0,
            m10: 0, m11: 1, m12: 0, m13: 0,
            m20: 0, m21: 0, m22: 1, m23: 0,
            m30: v.x, m31: v.y, m32: v.z, m33: 1
        )
    }

    static func scale(_ v: Vector3<T>) -> Self {
        Self(
            m00: v.x, m01: 0, m02: 0, m03: 0,
            m10: 0, m11: v.y, m12: 0, m13: 0,
            m20: 0, m21: 0, m22: v.z, m23: 0,
            m30: 0, m31: 0, m32: 0, m33: 1
        )
    }
}

extension Matrix4 where T: FloatingPoint & Mathable {
    static func * (left: Self, right: Self) -> Self {
        func dot(_ row: (T, T, T, T), _ col: (T, T, T, T)) -> T {
            return row.0 * col.0 + row.1 * col.1 + row.2 * col.2 + row.3 * col.3
        }
        return Self(
            m00: dot(
                (left.m00, left.m01, left.m02, left.m03),
                (right.m00, right.m10, right.m20, right.m30)),
            m01: dot(
                (left.m00, left.m01, left.m02, left.m03),
                (right.m01, right.m11, right.m21, right.m31)),
            m02: dot(
                (left.m00, left.m01, left.m02, left.m03),
                (right.m02, right.m12, right.m22, right.m32)),
            m03: dot(
                (left.m00, left.m01, left.m02, left.m03),
                (right.m03, right.m13, right.m23, right.m33)),
            m10: dot(
                (left.m10, left.m11, left.m12, left.m13),
                (right.m00, right.m10, right.m20, right.m30)),
            m11: dot(
                (left.m10, left.m11, left.m12, left.m13),
                (right.m01, right.m11, right.m21, right.m31)),
            m12: dot(
                (left.m10, left.m11, left.m12, left.m13),
                (right.m02, right.m12, right.m22, right.m32)),
            m13: dot(
                (left.m10, left.m11, left.m12, left.m13),
                (right.m03, right.m13, right.m23, right.m33)),
            m20: dot(
                (left.m20, left.m21, left.m22, left.m23),
                (right.m00, right.m10, right.m20, right.m30)),
            m21: dot(
                (left.m20, left.m21, left.m22, left.m23),
                (right.m01, right.m11, right.m21, right.m31)),
            m22: dot(
                (left.m20, left.m21, left.m22, left.m23),
                (right.m02, right.m12, right.m22, right.m32)),
            m23: dot(
                (left.m20, left.m21, left.m22, left.m23),
                (right.m03, right.m13, right.m23, right.m33)),
            m30: dot(
                (left.m30, left.m31, left.m32, left.m33),
                (right.m00, right.m10, right.m20, right.m30)),
            m31: dot(
                (left.m30, left.m31, left.m32, left.m33),
                (right.m01, right.m11, right.m21, right.m31)),
            m32: dot(
                (left.m30, left.m31, left.m32, left.m33),
                (right.m02, right.m12, right.m22, right.m32)),
            m33: dot(
                (left.m30, left.m31, left.m32, left.m33),
                (right.m03, right.m13, right.m23, right.m33))
        )
    }

    func translate(_ v: Vector3<T>) -> Self {
        Self.translation(v) * self
    }

    func scale(_ v: Vector3<T>) -> Self {
        Self.scale(v) * self
    }

    static func ortho(
        left: T,
        right: T,
        bottom: T,
        top: T,
        near: T,
        far: T
    ) -> Self {
        Self(
            m00: 2 / (right - left), m01: 0, m02: 0, m03: -(right + left) / (right - left),
            m10: 0, m11: 2 / (top - bottom), m12: 0, m13: -(top + bottom) / (top - bottom),
            m20: 0, m21: 0, m22: -2 / (far - near), m23: -(far + near) / (far - near),
            m30: 0, m31: 0, m32: 0, m33: 1
        )
    }
}

extension Matrix4 where T: FloatingPoint & Mathable & Sqrt & Trigonometry {
    static func rotation(axis: Vector3<T>, angle: Radians<T>) -> Self {
        let c = angle.cos
        let s = angle.sin
        let axis = axis.normalized
        return Self(
            m00: axis.x * axis.x * (1 - c) + c,
            m01: axis.x * axis.y * (1 - c) - axis.z * s,
            m02: axis.x * axis.z * (1 - c) + axis.y * s,
            m03: 0,
            m10: axis.y * axis.x * (1 - c) + axis.z * s,
            m11: axis.y * axis.y * (1 - c) + c,
            m12: axis.y * axis.z * (1 - c) - axis.x * s,
            m13: 0,
            m20: axis.x * axis.z * (1 - c) - axis.y * s,
            m21: axis.y * axis.z * (1 - c) + axis.x * s,
            m22: axis.z * axis.z * (1 - c) + c,
            m23: 0,
            m30: 0, m31: 0, m32: 0, m33: 1
        )
    }

    func rotate(axis: Vector3<T>, angle: Radians<T>) -> Self {
        Self.rotation(axis: axis, angle: angle) * self
    }

    static func perspective(
        verticalFieldOfView: Radians<T>,
        width: T,
        height: T,
        near: T,
        far: T
    ) -> Self {
        let f = 1 / (verticalFieldOfView / Radians(2)).tan
        let aspect = width / height
        return Self(
            m00: f / aspect, m01: 0, m02: 0, m03: 0,
            m10: 0, m11: f, m12: 0, m13: 0,
            m20: 0, m21: 0, m22: (far + near) / (near - far), m23: 2 * far * near / (near - far),
            m30: 0, m31: 0, m32: -1, m33: 0
        )
    }

    static func lookAt(
        position: Vector3<T>,
        target: Vector3<T>,
        up: Vector3<T>
    ) -> Self {
        let f = (target - position).normalized
        let up = up.normalized
        let s = Vector3.cross(f, up).normalized
        let u = Vector3.cross(s, f).normalized
        let mat = Self(
            m00: s.x, m01: u.x, m02: -f.x, m03: 0,
            m10: s.y, m11: u.y, m12: -f.y, m13: 0,
            m20: s.z, m21: u.z, m22: -f.z, m23: 0,
            m30: 0, m31: 0, m32: 0, m33: 1
        )
        return mat.translate(-position)
    }
}

extension Matrix4 where T: Mathable {
    func applyTo(vector: Vector3<T>) -> Vector3<T> {
        Vector3(
            x: m00 * vector.x + m10 * vector.y + m20 * vector.z,
            y: m01 * vector.x + m11 * vector.y + m21 * vector.z,
            z: m02 * vector.x + m12 * vector.y + m22 * vector.z
        )
    }

    func applyTo(point: Vector3<T>) -> Vector3<T> {
        Vector3(
            x: m00 * point.x + m10 * point.y + m20 * point.z + m30,
            y: m01 * point.x + m11 * point.y + m21 * point.z + m31,
            z: m02 * point.x + m12 * point.y + m22 * point.z + m32
        )
    }
}
